zilencer/views.py

```diff
@@ -500,6 +502,39 @@ def batch_create_table_data(
         row_objects = row_objects[BATCH_SIZE:]
 
 
+def update_remote_realm_data_for_server(
+    server: RemoteZulipServer, server_realms_info: List[Dict[str, Any]]
+) -> None:
+    if not server_realms_info:
+        return
+
+    uuids = [realm["uuid"] for realm in server_realms_info]
+    already_registered_uuids = [
+        str(uuid)
+        for uuid in RemoteRealm.objects.filter(uuid__in=uuids, server=server).values_list(
+            "uuid", flat=True
+        )
+    ]
+
+    new_remote_realms = [
+        RemoteRealm(
+            server=server,
+            uuid=realm["uuid"],
+            uuid_owner_secret=realm["uuid_owner_secret"],
+            host=realm["host"],
+            realm_deactivated=realm["deactivated"],
+            realm_date_created=timestamp_to_datetime(realm["date_created"]),
+        )
+        for realm in server_realms_info
+        if realm["uuid"] not in already_registered_uuids
+    ]
```

Comment:
I'm thinking about this right, this function should not only create new entries for new realms, but also look through all the `RemoteRealm` records and update **any** fields that may have changed?

Reply:
Yeah, I think that's correct. We probably want to create `RemoteRealmAuditLog` entries for those changes so we can keep track of them.

Reply:
Switched `already_registered_uuids` to a set just to not think about performance of that.

Reply:
Also switched to `uri` in the API; see https://github.com/zulip/zulip/issues/23380.

```diff
diff --git a/zerver/lib/remote_server.py b/zerver/lib/remote_server.py
index 22b526f948..5aad3419b3 100644
--- a/zerver/lib/remote_server.py
+++ b/zerver/lib/remote_server.py
@@ -180,7 +180,7 @@ def get_realms_info_for_push_bouncer() -> List[Dict[str, Any]]:
             uuid=str(realm.uuid),
             uuid_owner_secret=realm.uuid_owner_secret,
             host=realm.host,
-            uri=realm.uri,
+            url=realm.uri,
             deactivated=realm.deactivated,
             date_created=realm.date_created.timestamp(),
         )
diff --git a/zilencer/views.py b/zilencer/views.py
index 63bd8b0fea..bcf32bcfb1 100644
--- a/zilencer/views.py
+++ b/zilencer/views.py
@@ -507,12 +507,12 @@ def update_remote_realm_data_for_server(
     server: RemoteZulipServer, server_realms_info: List[Dict[str, Any]]
 ) -> None:
     uuids = [realm["uuid"] for realm in server_realms_info]
-    already_registered_uuids = [
+    already_registered_uuids = set(
         str(uuid)
         for uuid in RemoteRealm.objects.filter(uuid__in=uuids, server=server).values_list(
             "uuid", flat=True
         )
-    ]
+    )
 
     new_remote_realms = [
         RemoteRealm(
@@ -589,7 +589,7 @@ def remote_server_post_analytics(
                     ("uuid", check_string),
                     ("uuid_owner_secret", check_string),
                     ("host", check_string),
-                    ("uri", check_string),
+                    ("url", check_string),
                     ("deactivated", check_bool),
                     ("date_created", check_float),
                 ]
```